src/scf/newton.jl

```diff
@@ -0,0 +1,420 @@
+newton step δφ and the updated state φ - δφ
+(after proper orthonormalization) where δφ solves Jac * δφ = res
+and Jac is the Jacobian of the projected gradient descent : Jac = Ω+K.
+δφ is an element of the tangent space at φproj (set to φ if not specified in newton function).
+"""
+function newton_step(basis::PlaneWaveBasis, φ, φproj, res, occ;
+                     tol_krylov=1e-12, krylov_verbosity=1)
+
+    # necessary quantities
+    N = size(φ[1],2)
+    Nk = length(basis.kpoints)
+
+    # compute quantites at the point which define the tangent space
+    ρproj = compute_density(basis, φproj, occ)
+    energies, Hproj = energy_hamiltonian(basis, φproj, occ; ρ=ρproj)
+
+    # packing routines
+    pack(φ) = pack_arrays(basis, φ)
+    unpack(x) = unpack_arrays(basis, x)
+    packed_proj(δx, x) = pack(proj_tangent(unpack(δx), unpack(x)))
+
+    # mapping of the linear system on the tangent space
+    function f(x)
+        δφ = unpack(x)
+        δφ = proj_tangent(δφ, φproj)
+        Kδφ = apply_K(basis, δφ, φproj, ρproj, occ)
+        Ωδφ = apply_Ω(basis, δφ, φproj, Hproj)
+        pack(Ωδφ .+ Kδφ)
+    end
+
+    # project res on the good tangent space before starting
+    res = proj_tangent(res, φproj)
+
+    # solve (Ω+K) δφ = res on the tangent space with KrylovKit
+    δφ, info = linsolve(f, pack(res);
+                        tol=tol_krylov, verbosity=krylov_verbosity,
+                        # important to specify custom orthogonaliser to keep
+                        # iterations on the tangent space
+                        orth=OrthogonalizeAndProject(packed_proj, pack(φproj)))
+    δφ = unpack(δφ)
+    # ensure proper projection onto the tangent space
+    δφ = proj_tangent(δφ, φ)
+    φ_newton = similar(φ)
+
+    # perform newton_step
+    for ik = 1:Nk
+        φ_newton[ik] = ortho_qr(φ[ik] - δφ[ik])
+    end
+    (φ_newton, δφ)
+end
+
+"""
+    newton(basis::PlaneWaveBasis; ψ0=nothing,
+           tol=1e-6, max_iter=100, φproj=nothing, verbosity=0)
+
+Newton algorithm. Be careful that the starting needs to be not too far from the solution.
+Still contains debugging features.
+If φproj is nothing, we use the successive φ's to define the tangent spaces on which we solve
+Newton's equation. φproj can be useful when we want to fix the tangent space for testing.
+"""
+function newton(basis::PlaneWaveBasis{T}; ψ0=nothing,
+                tol=1e-6, max_iter=20, φproj=nothing, verbosity=0) where T
+
+    ## setting parameters
+    model = basis.model
+    @assert model.spin_polarization in (:none, :spinless, :collinear)
+    @assert model.temperature == 0 # temperature is not yet supported
+    filled_occ = filled_occupation(model)
+    n_spin = basis.model.n_spin_components
+    N = div(div(model.n_electrons, filled_occ), n_spin)
+
+    ## number of kpoints and occupation
+    Nk = length(basis.kpoints)
+    occupation = [filled_occ * ones(T, N) for ik = 1:Nk]
```

Comment:
I have some doubts here about the way we define occupation. I feel like we already know what the occupation is will be, right ?

Reply:
temperature is a lot more involved to do newton for so don't worry about that. I would just do the occupations like in direct_minimization

Reply:
Yes that's what I did, except that we need the step where we divide by `n_spin` to have the good size for occupation arrays. Or we completely forget about spins for the moment.

Reply:
I think direct_minimization does this correctly

Reply:
oh wait no it doesn't support spin.. Can you also update (and test) direct minimization with spins while you're at it?

Reply:
Yes ok, I'll do that, I keep you updated!

Reply:
Yes, works the same for `direct_minimization`, we just need to be careful when computing `n_bands` for the occupation.
```
n_spin = basis.model.n_spin_components                                       
n_bands = div(div(model.n_electrons, filled_occ), n_spin)
```

Reply:
But I still feel like it's not very useful if temperature is not supported...